src/Schemes/AffineSchemes/Objects/Attributes.jl

```diff
@@ -0,0 +1,275 @@
+
+# Examples
+```jldoctest
+julia> X = affine_space(QQ,3)
+Spec of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> codim(X)
+0
+```
+"""
+@attr function codim(X::Spec)
+  return dim(ideal(ambient_ring(X), [zero(ambient_ring(X))])) - dim(X)
+end
+
+
+@doc Markdown.doc"""
+    name(X::Spec)
+
+Returns the current name of an affine scheme.
+This name can be specified via `set_name!`.
+
+# Examples
+```jldoctest
+julia> X = affine_space(QQ, 3)
+Spec of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> name(X)
+"unnamed affine variety"
+
+julia> set_name!(X, "affine 3-dimensional space")
+
+julia> name(X)
+"affine 3-dimensional space"
+```
+"""
+@attr String function name(X::Spec)
+  return "unnamed affine variety"
+end
+
+
+function set_name!(X::AbsSpec, name::String)
+  return set_attribute!(X, :name, name)
+end
+
+
+########################################################################
+# (2) Further attributes for AbsSpec
+########################################################################
+
+# TODO: Needed?
+
+@Markdown.doc """
+    defining_ideal(X::AbsSpec{<:Any, <:MPolyRing})
+
+This method return the defining ideal of an affine scheme ``X``.
+
+# Examples
+```jldoctest
+julia> X = affine_space(QQ,3)
+Spec of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> R = OO(X)
+Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> (x1,x2,x3) = gens(R)
+3-element Vector{fmpq_mpoly}:
+ x1
+ x2
+ x3
+
+julia> Y = subscheme(X,ideal(R,[x1*x2]))
+Spec of Quotient of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field by ideal(x1*x2)
+
+julia> defining_ideal(Y)
+ideal(x1*x2)
+```
+"""
+@attr defining_ideal(X::AbsSpec{<:Any, <:MPolyRing}) = ideal(OO(X), [zero(OO(X))])
+defining_ideal(X::AbsSpec{<:Any, <:MPolyQuo}) = modulus(OO(X))
+@attr defining_ideal(X::AbsSpec{<:Any, <:MPolyLocalizedRing}) = ideal(OO(X), [zero(OO(X))])
+defining_ideal(X::AbsSpec{<:Any, <:MPolyQuoLocalizedRing}) = modulus(OO(X))
+
+
+@Markdown.doc """
+    strict_modulus(X::AbsSpec)
+
+This method return the strict modulus of an affine scheme ``X``.
+
+# Examples
+```jldoctest
+julia> X = affine_space(QQ,3)
+Spec of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> R = OO(X)
+Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> (x1,x2,x3) = gens(R)
+3-element Vector{fmpq_mpoly}:
+ x1
+ x2
+ x3
+
+julia> Y = subscheme(X,ideal(R,[x1*x2]))
+Spec of Quotient of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field by ideal(x1*x2)
+
+julia> strict_modulus(Y)
+ideal(x1*x2)
+```
+"""
+strict_modulus(X::AbsSpec) = saturated_ideal(modulus(OO(X)))
+
+
+
+########################################################################
+# (3) Further attributes for Spec
+########################################################################
+
+strict_modulus(X::Spec) = saturated_ideal(modulus(OO(X)))
```

Comment:
What is the difference between this an modulus(quotient_ring(OO(X)) (which can be cheaper)?

Reply:
I don't remember, to be honest. All this should be streamlined now.